File: api-country-code/src/main/java/com/kolyall/countryapi/api/CountryRepository.java
package com.kolyall.countryapi.api;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.RawRes;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.kolyall.countryapi.objects.Country;
import com.kolyall.countryapi.objects.Language;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.reflect.Type;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.Callable;

import rx.Observable;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

import static android.content.ContentValues.TAG;

/**
 * Created by Nick Unuchek on 11.09.2017.
 */

public class CountryRepository {

    Context mContext;
    Gson mGson;

    public CountryRepository(Context context, Gson gson) {
        mContext = context;
        mGson = gson;
    }

    public Observable<List<Country>> getCountryListObservable(){
        Type type = new TypeToken<List<Country>>(){}.getType();
        return getResponseFromRes(getLanguage().getRawJsonId(), (Type) type);
    }

    public Observable<Country> getCountryByRegionCodeObservable(String regionCode){
        return getCountryListObservable()
                .flatMap(Observable::from)
                .filter(country -> country.getRegionCode().equals(regionCode.toLowerCase()));
    }

    private Language getLanguage() {
        String localeLanguage = Locale.getDefault().getLanguage();
        for (Language language:Language.values()){
            if (language.getCode().equals(localeLanguage)){
                return language;
            }
        }
        return Language.ENGLISH;
    }

    @NonNull
    protected <T> Observable<T> getResponseFromRes(@RawRes int rawId, final Type type) {
        return readStringFromResObservable(rawId)
                .map(new Func1<String, T>() {
                    @Override
                    public T call(String jsonObject) {
                        Log.d(TAG, "call: getResponseFromRes: map: "+Thread.currentThread().getName());
                        return jsonToObj(jsonObject, type);
                    }
                });
    }

    @NonNull
    private Observable<String> readStringFromResObservable(@RawRes final int rawId) {
        return Observable.fromCallable(new Callable<String>() {
            @Override
            public String call() throws Exception {
                Log.d(TAG, "call: readStringFromResObservable fromCallable: "+Thread.currentThread().getName());
                return readJsonFromRes(rawId);
            }
        }).subscribeOn(Schedulers.io());
    }

    private String readJsonFromRes(int rawId) throws IOException {
        Log.d(TAG, "call: readJsonFromRes method: "+Thread.currentThread().getName());
        InputStream is = mContext.getResources().openRawResource(rawId);
        Writer writer = new StringWriter();
        char[] buffer = new char[1024];
        try {
            Reader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
            int n;
            while ((n = reader.read(buffer)) != -1) {
                writer.write(buffer, 0, n);
            }
        } finally {
            is.close();
        }

        return writer.toString();
    }

    private <T> T jsonToObj(String json,Type type) {
        return mGson.fromJson(json, type);
    }

}

File: api-country-code/src/main/java/com/kolyall/countryapi/api/CountryRepository.kt
package com.kolyall.countryapi.api

import android.content.ContentValues
import android.content.Context
import android.util.Log
import androidx.annotation.RawRes
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.kolyall.countryapi.objects.Country
import com.kolyall.countryapi.objects.Language
import io.reactivex.Maybe
import io.reactivex.schedulers.Schedulers
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader
import java.io.StringWriter
import java.io.Writer
import java.lang.reflect.Type
import java.util.Locale

/**
 * Created by Nick Unuchek on 11.09.2017.
 */
class CountryRepository(private var context: Context, private var gson: Gson) {

    fun getCountryList(localeLanguageCode: String = Locale.getDefault().language): Maybe<List<Country>> {
        val type = object : TypeToken<List<Country>>() {}.type
        return getResponseFromRes(getLanguage(localeLanguageCode).rawJsonId, type as Type)
    }

    fun getCountryByRegionCode(regionCode: String, localeLanguageCode: String = Locale.getDefault().language): Maybe<Country> {
        return getCountryList(localeLanguageCode)
            .map { list ->
                list.firstOrNull { country -> country.regionCode == regionCode.toLowerCase() }
            }
    }

    private fun getLanguage(localeLanguageCode: String): Language {
        for (language in Language.values()) {
            if (language.code == localeLanguageCode) {
                return language
            }
        }
        return Language.ENGLISH
    }

    protected fun <T> getResponseFromRes(@RawRes rawId: Int, type: Type): Maybe<T> {
        return readStringFromResObservable(rawId)
            .map { jsonObject ->
                Log.d(
                    ContentValues.TAG,
                    "call: getResponseFromRes: map: " + Thread.currentThread().name
                )
                jsonToObj<T>(jsonObject, type)
            }
    }

    private fun readStringFromResObservable(@RawRes rawId: Int): Maybe<String> {
        return Maybe.fromCallable {
            Log.d(
                ContentValues.TAG,
                "call: readStringFromResObservable fromCallable: " + Thread.currentThread().name
            )
            readJsonFromRes(rawId)
        }.subscribeOn(Schedulers.io())
    }

    @Throws(IOException::class)
    private fun readJsonFromRes(rawId: Int): String {
        Log.d(ContentValues.TAG, "call: readJsonFromRes method: " + Thread.currentThread().name)
        val stream = context.resources.openRawResource(rawId)
        val writer: Writer = StringWriter()
        val buffer = CharArray(1024)
        stream.use { useStream ->
            val reader: Reader = BufferedReader(InputStreamReader(useStream, Charsets.UTF_8))
            var n: Int
            while (reader.read(buffer).also { n = it } != -1) {
                writer.write(buffer, 0, n)
            }
        }
        return writer.toString()
    }

    private fun <T> jsonToObj(json: String, type: Type): T {
        return gson.fromJson(json, type)
    }
}